src/steamship/plugin/capabilities.py

```diff
@@ -0,0 +1,291 @@
+        :return: a bool that indicates whether the Plugin support level is valid, plus an Optional Response for this
+          capability's support, if applicable.
+        """
+        if self.request_level == RequestLevel.OPTIONAL and support_level is None:
+            # We don't support it, but it's optional.
+            return True, None
+        if support_level is None:
+            # Not optional, but we don't support it.
+            return False, None
+        if self.request_level == RequestLevel.NATIVE and support_level != RequestLevel.NATIVE:
+            # They want native support, but we don't support it at a native level.
+            return False, None
+        # They want NATIVE support, and we support that, OR
+        # They want BEST_EFFORT support, and we support that, OR
+        # They want BEST_EFFORT and we support native
+        return True, Capability.Response(fulfilled_at=support_level)
+
+    class Response(BaseModel, extra=Extra.allow):
+        """Response regarding a specific capability served by the plugin.
+
+        Responses indicate at minimum the level at which they served the request-level.  They may
+        also specify additional metadata.
+        """
+
+        CAPABILITY_NAME: ClassVar[str]
+
+        name: str = Field(init_var=False, default=None)
+        fulfilled_at: RequestLevel
+
+
+class CapabilityImpl(Capability, extra=Extra.forbid):
+    # TODO (PR): The Extra.forbid here is to enforce clamping of deserialization, but it may just be simpler to leave
+    #  that up to individual capabilities?  My goal here is to prevent accidental breakage of contract when someone
+    #  provides specific metadata that makes it so e.g. a plugin with an older view of the world thinks it has Best
+    #  Effort support but can't because of those extra requests.
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.name = self.NAME
+
+    class ResponseImpl(Capability.Response, extra=Extra.forbid):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.name = self.CAPABILITY_NAME
+
+
+class UnsupportedCapabilityError(Exception):
+    """
+    Raised when support for a given feature is requested from a plugin in a non-optional way, but the plugin doesn't
+    support it
+    """
+
+    def __init__(self, unsupported_capabilities: List[Capability]):
+        names = ", ".join([cap.name for cap in unsupported_capabilities])
+        super().__init__(
+            f"The following features were requested but are not supported by this plugin: {names}"
+        )
+        self.unsupported_capabilities = unsupported_capabilities
+
+
+class CapabilityPluginRequest(BaseModel):
+    """Model representing the text in a STEAMSHIP_PLUGIN_CAPABILITIES block when it is used as a request."""
+
+    # TODO (PR): Might be cleaner to have separate MIMEs for response/request.
```

Comment:
Assuming the MIME/block approach I have here makes sense, I'm just going to do this for clarity.